data-plane/receiver-loom/src/main/java/dev/knative/eventing/kafka/broker/receiverloom/LoomKafkaProducer.java

```diff
@@ -15,6 +15,146 @@
+        Promise<RecordMetadata> promise = Promise.promise();
+        if (isClosed.get()) {
+            promise.fail("Producer is closed");
+        } else {
+            eventQueue.add(new RecordPromise(record, promise));
+        }
+        return promise.future();
+    }
+
+    private void sendFromQueue() {
+        while (!Thread.currentThread().isInterrupted()) {
+            try {
+                RecordPromise recordPromise = eventQueue.take();
+                ProducerTracer.StartedSpan startedSpan =
+                        this.tracer == null ? null : this.tracer.prepareSendMessage(ctx, recordPromise.getRecord());
+                producer.send(recordPromise.getRecord(), (metadata, exception) -> {
+                    if (exception != null) {
+                        recordPromise.getPromise().fail(exception);
+                        if (startedSpan != null) {
+                            startedSpan.fail(ctx, exception);
+                        }
+                    }
+                    if (startedSpan != null) {
+                        startedSpan.finish(ctx);
+                    }
+                    recordPromise.getPromise().complete(metadata);
+                });
+            } catch (InterruptedException e) {
+                logger.debug("Interrupted while waiting for event queue to be populated.");
+            }
+        }
+    }
+
+    @Override
+    public Future<Void> close() {
+        Promise<Void> promise = Promise.promise();
+        this.isClosed.set(true);
+        Thread.ofVirtual().start(() -> {
+            try {
+                while (!eventQueue.isEmpty()) {
+                    Thread.sleep(2000L);
+                }
+                sendFromQueueThread.interrupt();
+                producer.close();
+                sendFromQueueThread.join();
+                promise.complete();
+            } catch (Exception e) {
+                promise.fail(e);
+            }
+        });
+        return promise.future();
+    }
+
+    @Override
+    public Future<Void> flush() {
+        Promise<Void> promise = Promise.promise();
+        Thread.ofVirtual().start(() -> {
+            try {
+                producer.flush();
+                promise.complete();
+            } catch (Exception e) {
+                promise.fail(e);
+            }
+        });
+        return promise.future();
+    }
+
+    @Override
+    public Producer<K, V> unwrap() {
+        return producer;
+    }
+
+    private class RecordPromise {
+        private final ProducerRecord<K, V> record;
+        private final Promise<RecordMetadata> promise;
+
+        private RecordPromise(ProducerRecord<K, V> record, Promise<RecordMetadata> promise) {
+            this.record = record;
+            this.promise = promise;
+        }
+
+        public ProducerRecord<K, V> getRecord() {
+            return record;
+        }
+
+        public Promise<RecordMetadata> getPromise() {
+            return promise;
+        }
+    }
```

Comment:
I would rework this class to be static and defined on generic `<K, V>`

```suggestion
    private static class RecordPromise<K, V> {
        private final ProducerRecord<K, V> record;
        private final Promise<RecordMetadata> promise;

        private RecordPromise(ProducerRecord<K, V> record, Promise<RecordMetadata> promise) {
            this.record = record;
            this.promise = promise;
        }

        public ProducerRecord<K, V> getRecord() {
            return record;
        }

        public Promise<RecordMetadata> getPromise() {
            return promise;
        }
    }
```

To reduce the risk of the introduction (now and in the future) of any form of memory leak https://www.infoworld.com/article/3526554/avoid-memory-leaks-in-inner-classes.html

Reply:
that's a good call. I am also in fav. of this